src/order/heyting/regular.lean

```diff
@@ -0,0 +1,150 @@
+instance : has_compl (heyting_regular α) := ⟨λ a, ⟨aᶜ, is_heyting_regular_compl _⟩⟩
+
+@[simp, norm_cast] lemma coe_top : ((⊤ : heyting_regular α) : α) = ⊤ := rfl
+@[simp, norm_cast] lemma coe_bot : ((⊥ : heyting_regular α) : α) = ⊥ := rfl
+@[simp, norm_cast] lemma coe_inf (a b : heyting_regular α) : (↑(a ⊓ b) : α) = a ⊓ b := rfl
+@[simp, norm_cast] lemma coe_himp (a b : heyting_regular α) : (↑(a ⇨ b) : α) = a ⇨ b := rfl
+@[simp, norm_cast] lemma coe_compl (a : heyting_regular α) : (↑(aᶜ) : α) = aᶜ := rfl
+
+instance : inhabited (heyting_regular α) := ⟨⊥⟩
+instance : semilattice_inf (heyting_regular α) := coe_injective.semilattice_inf _ coe_inf
+instance : bounded_order (heyting_regular α) := bounded_order.lift coe (λ _ _, id) coe_top coe_bot
+
+@[simp, norm_cast] lemma coe_le_coe {a b : heyting_regular α} : (a : α) ≤ b ↔ a ≤ b := iff.rfl
+@[simp, norm_cast] lemma coe_lt_coe {a b : heyting_regular α} : (a : α) < b ↔ a < b := iff.rfl
+
+/-- The smallest regular element greater than `a`. -/
+def to_heyting_regular : α →o heyting_regular α :=
+⟨λ a, ⟨aᶜᶜ, is_heyting_regular_compl _⟩, λ a b h, coe_le_coe.1 $ compl_le_compl $ compl_le_compl h⟩
+
+@[simp, norm_cast] lemma coe_to_heyting_regular (a : α) : (to_heyting_regular a : α) = aᶜᶜ := rfl
+@[simp] lemma to_heyting_regular_coe (a : heyting_regular α) : to_heyting_regular (a : α) = a :=
+coe_injective a.2
+
+/-- The Galois insertion between `heyting_regular.to_heyting_regular` and `coe`. -/
+def gi : galois_insertion to_heyting_regular (coe : heyting_regular α → α) :=
+{ choice := λ a ha, ⟨a, ha.antisymm le_compl_compl⟩,
+  gc := λ a b, coe_le_coe.symm.trans $
+    ⟨le_compl_compl.trans, λ h, (compl_anti $ compl_anti h).trans_eq b.2⟩,
+  le_l_u := λ _, le_compl_compl,
+  choice_eq := λ a ha, coe_injective $ le_compl_compl.antisymm ha }
+
+instance : lattice (heyting_regular α) := gi.lift_lattice
+
+@[simp, norm_cast] lemma coe_sup (a b : heyting_regular α) : (↑(a ⊔ b) : α) = (a ⊔ b)ᶜᶜ := rfl
+
+instance : boolean_algebra (heyting_regular α) :=
+{ le_sup_inf := λ a b c, coe_le_coe.1 $ by { dsimp, rw [sup_inf_left, compl_compl_inf_distrib] },
+  inf_compl_le_bot := λ a, coe_le_coe.1 disjoint_compl_right,
+  top_le_sup_compl := λ a, coe_le_coe.1 $
+    by { dsimp, rw [compl_sup, inf_compl_eq_bot, compl_bot], refl },
+  himp_eq := λ a b, coe_injective begin
+    dsimp,
+    rw [compl_sup, a.prop.eq],
+    refine eq_of_forall_le_iff (λ c, le_himp_iff.trans _),
+    rw [le_compl_iff_disjoint_right, disjoint_left_comm, b.prop.disjoint_compl_left_iff],
+  end,
+  ..heyting_regular.lattice, ..heyting_regular.bounded_order, ..heyting_regular.has_himp,
+  ..heyting_regular.has_compl }
+
+@[simp, norm_cast] lemma coe_sdiff (a b : heyting_regular α) : (↑(a \ b) : α) = a ⊓ bᶜ := rfl
+
+end heyting_regular
+end heyting_algebra
+
+variables [boolean_algebra α]
+
+lemma is_heyting_regular_of_boolean : ∀ a : α, is_heyting_regular a := compl_compl
```

Comment:
Do we want to prove that `decidable P → is_heyting_regular P` in `Prop`?

Reply:
Weird stuff happening here.
```
lemma is_heyting_regular_of_decidable (p : Prop) [decidable p] : is_heyting_regular p :=
propext $ decidable.not_not_iff _

#print axioms is_heyting_regular_of_decidable

/-
propext
classical.choice
quot.sound
-/
```
I checked everything that appeared in the trace of `#print is_heyting_regular_of_decidable` and the most it uses is `propext`.

Reply:
It fixed itself now. No idea why.